### PR TITLE
Question 2: Maximum Subarray

### PR DESCRIPTION
**Pull request by RyanRussell00**

### CODE REVIEW

File: 2020-02-Spring/0_arrays_matrix/README.md [hunk 1]
```diff
@@ -119,21 +125,43 @@ Go to [Top](#top)
...
-#### Optimal Solution
-
-TODO :bug:
-
-#### Testing The Solutions OR Driver For Solution
-
-TODO :bug:
+### 2. SOLUTION 2
+
+Source: https://leetcode.com/problems/maximum-subarray/
```
Comment (etcadinfinitum) [inline 2020-02-Spring/0_arrays_matrix/README.md:130]: Markdown link embedding is nice.

```suggestion
Source: [LeetCode](https://leetcode.com/problems/maximum-subarray/)
```

File: 2020-02-Spring/0_arrays_matrix/maximum-subarray/maxSubArray.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
+class Solution {
```
Comment (etcadinfinitum) [inline 2020-02-Spring/0_arrays_matrix/maximum-subarray/maxSubArray.java:1]: Could use some tests here... if you have time.

Also, if I were to clone this and compile the class as-is, I wouldn't be able to due to the discrepancy between the file name and the class name. (Code that doesn't work out of the box is annoying and not the best look.)

File: 2020-02-Spring/0_arrays_matrix/README.md [hunk 3]
```diff
@@ -50,19 +50,25 @@ Go to [Solution](#s1)   [Top](#top)
 
 ### 2. PROBLEM 2 TODO :bug:
```
Comment (anishprasad01) [inline 2020-02-Spring/0_arrays_matrix/README.md:51]: Need the name of the problem and difficulty
```suggestion
### 2. Maximum Subarray (Easy)
```

File: 2020-02-Spring/0_arrays_matrix/README.md [hunk 4]
```diff
@@ -50,19 +50,25 @@ Go to [Solution](#s1)   [Top](#top)
...
 
-Source: TODO :bug:
+Source: https://leetcode.com/problems/maximum-subarray/
 
 #### Scenario
 
-Problem Statement TODO :bug:
+Given an integer array nums, find the contiguous subarray (containing at least one number) which has the largest sum and return its sum.
```
Comment (anishprasad01) [inline 2020-02-Spring/0_arrays_matrix/README.md:57]: I'm adding a small change here, but the explanation needs a little more detail. Things like what is considered correct input, how to format output, whether or not correct input is guaranteed, etc.

```suggestion
Given an array of integers, find the contiguous subarray (containing at least one number) which has the largest sum and return the sum.
```

File: 2020-02-Spring/0_arrays_matrix/README.md [hunk 5]
```diff
@@ -119,21 +125,43 @@ Go to [Top](#top)
...
+So if we are looking at this array: `[-2,1,-3,4,-1,2,1,-5,4]`
+We will look at -2 and see we have a negative number, so set curr to 0. 
+Then we see that 1 is postive, set curr & max to 1. 
+Then -3 gives us a curr of -2, set curr to 0, but the max is still 1.
+Then 4 gives us a curr of 4, and update max to 4. Since our curr is now 4, we have ignored all the values before 4 (-2, 1, -3).
+
+#### Solution
+
+```java
+public int maxSubArray(int[] nums) {
```
Comment (etcadinfinitum) [inline 2020-02-Spring/0_arrays_matrix/README.md:144]: Code style for this is good. Short, sweet, and highly readable. (I use the same bracket style so it's definitely up my alley, but I appreciate the judicious choice of variable names etc as well.) :tada: 

I traced through a few examples by hand. This solution appears to work for the edge cases I came up with, but I strongly encourage doing the following:
1. Writing tests to prove it works.
2. Tracing through edge cases with this code during the session (or even giving example input to let folks test their own solutions by and or in an editor).

Here are a few edge cases I tried:
* All negative numbers (`[ -3 -16 -4 -1 -2 -6]`)
* Not front-loaded arrays, where the max is at the end of the array (`[ -1 1 3 4 5 ]`)
* A normal-ish array (`[ 3 2 6 0 1 -3 1 ]`)

File: 2020-02-Spring/0_arrays_matrix/README.md [hunk 6]
```diff
@@ -119,21 +125,43 @@ Go to [Top](#top)
...
-
-TODO :bug:
+### 2. SOLUTION 2
+
+Source: https://leetcode.com/problems/maximum-subarray/
+
+We will traverse over the array and keep track of 2 variables: Total maximum sum, and the current sum.
+We will look at each number and add it to the current sum, if the current sum becomes negative, we ignore the numbers that made it negative.
+Resetting the current to 0 is basically saying "Ignore everything we looked at previously because it's not getting us the greatest sum."
+So if we are looking at this array: `[-2,1,-3,4,-1,2,1,-5,4]`
```
Comment (etcadinfinitum) [inline 2020-02-Spring/0_arrays_matrix/README.md:135]: This renders as one large paragraph; recommend breaking it up into explanation + example.

large_paragraph


```suggestion

So if we are looking at this array: `[-2,1,-3,4,-1,2,1,-5,4]`
```

File: 2020-02-Spring/0_arrays_matrix/README.md [hunk 7]
```diff
@@ -119,21 +125,43 @@ Go to [Top](#top)
...
-TODO :bug:
-
-#### Optimal Solution
-
-TODO :bug:
-
-#### Testing The Solutions OR Driver For Solution
-
-TODO :bug:
+### 2. SOLUTION 2
```
Comment (anishprasad01) [inline 2020-02-Spring/0_arrays_matrix/README.md:128]: Need problem name
```suggestion
### 2. Maximum Subarray Solution
```

File: 2020-02-Spring/0_arrays_matrix/README.md [hunk 8]
```diff
@@ -119,21 +125,43 @@ Go to [Top](#top)
...
+We will traverse over the array and keep track of 2 variables: Total maximum sum, and the current sum.
+We will look at each number and add it to the current sum, if the current sum becomes negative, we ignore the numbers that made it negative.
+Resetting the current to 0 is basically saying "Ignore everything we looked at previously because it's not getting us the greatest sum."
+So if we are looking at this array: `[-2,1,-3,4,-1,2,1,-5,4]`
+We will look at -2 and see we have a negative number, so set curr to 0. 
+Then we see that 1 is postive, set curr & max to 1. 
+Then -3 gives us a curr of -2, set curr to 0, but the max is still 1.
+Then 4 gives us a curr of 4, and update max to 4. Since our curr is now 4, we have ignored all the values before 4 (-2, 1, -3).
+
+#### Solution
```
Comment (anishprasad01) [inline 2020-02-Spring/0_arrays_matrix/README.md:141]: Need a solution source too

File: 2020-02-Spring/0_arrays_matrix/README.md [hunk 9]
```diff
@@ -119,21 +125,43 @@ Go to [Top](#top)
...
+    //Keep track of the max sum seen so far, and the current running sum.
+    int max = nums[0];
+    int curr = 0;
+    for (int i = 0; i < nums.length; i++) {
+        curr += nums[i];
+        //If the current running sum is greater than the total, set it to max.
+        if (curr > max) {
+            max = curr;
+        }
+        //If the current running total becomes negative, we ignore the numbers
```
Comment (anishprasad01) [inline 2020-02-Spring/0_arrays_matrix/README.md:157]: Ignore by itself sounds weird. Suggest rewording it.
```suggestion
        //If the current running total becomes negative, we ignore all previous numbers
```

File: 2020-02-Spring/0_arrays_matrix/maximum-subarray/maxSubArray.java [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+    //Keep track of the max sum seen so far, and the current running sum.
+    int max = nums[0];
+    int curr = 0;
+    for (int i = 0; i < nums.length; i++) {
+        curr += nums[i];
+        //If the current running sum is greater than the total, set it to max.
+        if (curr > max) {
+            max = curr;
+        }
+        //If the current running total becomes negative, we ignore the numbers
```
Comment (anishprasad01) [inline 2020-02-Spring/0_arrays_matrix/maximum-subarray/maxSubArray.java:15]: Same thing with the "ignore" as in the doc
```suggestion
        //If the current running total becomes negative, we ignore all previous numbers
```

File: 2020-02-Spring/0_arrays_matrix/README.md [hunk 11]
```diff
@@ -119,21 +125,43 @@ Go to [Top](#top)
...
+
+We will traverse over the array and keep track of 2 variables: Total maximum sum, and the current sum.
+We will look at each number and add it to the current sum, if the current sum becomes negative, we ignore the numbers that made it negative.
+Resetting the current to 0 is basically saying "Ignore everything we looked at previously because it's not getting us the greatest sum."
+So if we are looking at this array: `[-2,1,-3,4,-1,2,1,-5,4]`
+We will look at -2 and see we have a negative number, so set curr to 0. 
+Then we see that 1 is postive, set curr & max to 1. 
+Then -3 gives us a curr of -2, set curr to 0, but the max is still 1.
+Then 4 gives us a curr of 4, and update max to 4. Since our curr is now 4, we have ignored all the values before 4 (-2, 1, -3).
+
```
Comment (anishprasad01) [inline 2020-02-Spring/0_arrays_matrix/README.md:140]: Might want to finish this explanation. You've stopped in the middle of traversing this array.

File: 2020-02-Spring/0_arrays_matrix/README.md (hunk 5 above)
Comment (RyanRussell00) [reply 2020-02-Spring/0_arrays_matrix/README.md:144]: Will add those, thank you!

File: 2020-02-Spring/0_arrays_matrix/README.md (hunk 8 above)
Comment (RyanRussell00) [reply 2020-02-Spring/0_arrays_matrix/README.md:141]: Solution is in the other file

File: 2020-02-Spring/0_arrays_matrix/README.md (hunk 9 above)
Comment (RyanRussell00) [reply 2020-02-Spring/0_arrays_matrix/README.md:157]: Fixed

File: 2020-02-Spring/0_arrays_matrix/README.md (hunk 11 above)
Comment (RyanRussell00) [reply 2020-02-Spring/0_arrays_matrix/README.md:140]: Fixed

File: 2020-02-Spring/0_arrays_matrix/README.md (hunk 8 above)
Comment (etcadinfinitum) [reply 2020-02-Spring/0_arrays_matrix/README.md:146]: > Need a solution source too

Not sure what you're asking here but here's my 2 cents:
* I don't think I've ever given credit attribution to a solution before, mostly because I would typically write them myself (LC especially) or use the solution in the CTCI book.
* There is a `Solution.java` in this commit

File: 2020-02-Spring/0_arrays_matrix/maximum-subarray/maxSubArray.java [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+    //Keep track of the max sum seen so far, and the current running sum.
+    int max = nums[0];
+    int curr = 0;
+    for (int i = 0; i < nums.length; i++) {
+        curr += nums[i];
+        //If the current running sum is greater than the total, set it to max.
+        if (curr > max) {
+            max = curr;
+        }
+    //If the current running total becomes negative, we ignore all previous numbers        if (curr < 0) {
```
Comment (etcadinfinitum) [inline 2020-02-Spring/0_arrays_matrix/maximum-subarray/maxSubArray.java:14]: Whoops, the if statement was commented out
```suggestion
        //If the current running total becomes negative, we ignore all previous numbers
        if (curr < 0) {
```

File: 2020-02-Spring/0_arrays_matrix/README.md (hunk 1 above)
Comment (etcadinfinitum) [reply 2020-02-Spring/0_arrays_matrix/README.md:130]: This is still nice

File: 2020-02-Spring/0_arrays_matrix/README.md (hunk 6 above)
Comment (etcadinfinitum) [reply 2020-02-Spring/0_arrays_matrix/README.md:137]: You should still do this

File: 2020-02-Spring/0_arrays_matrix/maximum-subarray/maxSubArray.java [hunk 13]
```diff
@@ -0,0 +1,18 @@
+public int maxSubArray(int[] nums) {
```
Comment (etcadinfinitum) [inline 2020-02-Spring/0_arrays_matrix/maximum-subarray/maxSubArray.java:1]: This will not compile because a class declaration is missing. Wrap the method in a class declaration, and ensure the class name matches the file name.

I know you've been living in c++ land for the past year or so, but if you're going to write Java, make sure it compiles :+1:

File: 2020-02-Spring/0_arrays_matrix/README.md (hunk 8 above)
Comment (anishprasad01) [reply 2020-02-Spring/0_arrays_matrix/README.md:146]: Oh, I thought we needed one for some reason. Never mind then.

File: 2020-02-Spring/0_arrays_matrix/README.md [hunk 14]
```diff
@@ -50,19 +50,25 @@ Go to [Solution](#s1)   [Top](#top)
 
 ### 2. PROBLEM 2 TODO :bug:
 
-Source: TODO :bug:
+Source: https://leetcode.com/problems/maximum-subarray/
```
[the remark this answers is not in the document]
Comment (RyanRussell00) [reply 2020-02-Spring/0_arrays_matrix/README.md:53]: Going to avoid embedded links because I've had issues where the permalinks doesn't work. They could have been an ID10T error, but I'm gonna play it safe than sorry.

File: 2020-02-Spring/0_arrays_matrix/README.md (hunk 6 above)
Comment (RyanRussell00) [reply 2020-02-Spring/0_arrays_matrix/README.md:137]: Fixed in next commit

File: 2020-02-Spring/0_arrays_matrix/maximum-subarray/maxSubArray.java (hunk 13 above)
Comment (RyanRussell00) [reply 2020-02-Spring/0_arrays_matrix/maximum-subarray/maxSubArray.java:1]: Fixed in next commit

File: 2020-02-Spring/0_arrays_matrix/maximum-subarray/maxSubArray.java [hunk 15]
```diff
@@ -1,5 +1,9 @@
+<<<<<<< HEAD
 class maxSubArray {
     public int maxSubArray(int[] nums) {
+=======
+public int maxSubArray(int[] nums) {
+>>>>>>> b414c690747c14860bd950be536b7b77faf30ddf
```
Comment (anishprasad01) [inline 2020-02-Spring/0_arrays_matrix/maximum-subarray/maxSubArray.java:6]: You have some merge syntax left over.
